### PR TITLE
adding a list users API endpoint, with autocomplete filtering

### PR DESCRIPTION
**Pull request by sekharpanja**
Added a new API endpoint for applications to be able to search for users, with autocomplete. By default, search is restricted to those with access to the app that is searching.

### CODE REVIEW

File: sso/user/views.py [hunk 1]
```diff
@@ -63,3 +77,82 @@ def retrieve(self, request):
...
+            # check if application is in permitted_applications
+            permitted_qs = queryset.filter(permitted_applications=application)
+            # check if application is in access_profiles's apps
+            access_qs = queryset.filter(access_profiles__oauth2_applications=application)
+            if email_qs:
+                qs = email_qs | permitted_qs | access_qs
+            else:
+                qs = permitted_qs | access_qs
+            return qs.distinct()    # remove dups
+        else:
```
Comment (lgarvey) [inline sso/user/views.py]: I think it should technically be impossible to get here.  And if it were possible, I expect that it would blow up on line 113 first, as I think `request.user.auth.application` is set by django-oauth-toolkit.

File: sso/user/views.py [hunk 2]
```diff
@@ -63,3 +77,82 @@ def retrieve(self, request):
...
+
+    def filter_queryset(self, queryset):
+        """
+        Applies the default ordering when the query set has not already been ordered.
+
+        (The autocomplete filter automatically applies an ordering, hence we only set the
+        default ordering when another one has not already been set.)
+        """
+        filtered_queryset = super().filter_queryset(queryset)
+
```
Comment (lgarvey) [inline sso/user/views.py]: nit picking - double space

File: sso/user/views.py [hunk 3]
```diff
@@ -63,3 +77,82 @@ def retrieve(self, request):
...
+                qs = email_qs | permitted_qs | access_qs
+            else:
+                qs = permitted_qs | access_qs
+            return qs.distinct()    # remove dups
+        else:
+            # do we ever get into this?
+            return queryset.filter(
+                Q(access_profiles__saml2_applications__enabled=True) & \
+                Q(access_profiles__saml2_applications=application)
+            )
```
Comment (lgarvey) [inline sso/user/views.py]: is any sort of pagination required?  With 'default_access_allowed` we could be looking at nearly 6k results.

File: sso/user/serializers.py [hunk 4]
```diff
@@ -55,3 +55,24 @@ def validate(self, data):
...
+            'user_id',
+            'first_name',
+            'last_name',
+            'email',
+        )
+
+    def to_representation(self, obj):
+        app = self.context['request'].auth.application
+
+        primary_email, related_emails = obj.get_emails_for_application(app)
```
Comment (lgarvey) [inline sso/user/serializers.py]: do you need related emails here? If not, it's probalby worth being explicit using `_`

File: .gitignore [hunk 5]
```diff
@@ -24,3 +24,5 @@ __py_cache__/
 .pytest_cache/
 
 /config/google_credentials.json
+
+.vscode/
```
Comment (mcbhenwood) [inline .gitignore]: Updating .gitignore should be in a separate commit. (Also, arguably, something as dev-specific as the choice of code editor (and the ignorance of its local files) should probably go into the developer's own `.gitignore` file.

File: sso/user/serializers.py [hunk 6]
```diff
@@ -68,7 +68,7 @@ class Meta:
     def to_representation(self, obj):
         app = self.context['request'].auth.application
 
-        primary_email, related_emails = obj.get_emails_for_application(app)
+        primary_email, _ = obj.get_emails_for_application(app)
```
Comment (mcbhenwood) [inline sso/user/serializers.py]: Rather than dummy variable `_` would it be preferable to write this?
```python
primary_email = obj.get_emails_for_application(app)[0]
```

File: .gitignore (hunk 5 above)
Comment (sekharpanja) [reply .gitignore]: Sure, Mark. Good point.

File: sso/user/serializers.py (hunk 6 above)
Comment (sekharpanja) [reply sso/user/serializers.py]: Dummy variable is generally my preference as well, unless there is a strong reason against it.

File: sso/user/views.py [hunk 7]
```diff
@@ -63,3 +77,77 @@ def retrieve(self, request):
...
+    )
+    filterset_class = UserAutoCompleteFilter
+    ordering_fields = ('first_name', 'last_name')
+    _default_ordering = ('first_name', 'last_name')
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        application = self.request.auth.application
+
+        if isinstance(application, OAuthApplication):
```
Comment (currycoder) [inline sso/user/views.py]: Could this be:
`if application:` instead?  I'm not quite sure on the contract, but this had me wondering what other type application could be that was allowable.  If it's just going to be None or falsey, this could reduce some ambiguity.

Comment (currycoder) [inline sso/user/views.py]: I would be tempted to make this whole block either a helper on the viewset or a queryset/manager method for the user model.

It could make things a bit more self documenting, reusable and testable if it was present in a separate method somewhere.

File: sso/user/autocomplete.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            'pk',
+        )
+
+    filter_q_objects_for_tokens = (
+        _make_filter_q_for_token(autocomplete_fields, escaped_token)
+        for escaped_token in escaped_tokens
+    )
+
+    return queryset.annotate(
+        _matched_group_index=_make_ordering_case_expression(
```
Comment (currycoder) [inline sso/user/autocomplete.py]: This method is very readable IMO until we get to the `_make_ordering_case_expression` parts.  For me these bits add a fair bit of complexity that make a developer need to drill in to their helpers to fully understand.  (Even now after a few re-reads I don't have a confident understanding to be able to maintain).  I do like that they give some determinism to the ordering, but what do we really gain from this?

My thought is that a user will likely find the result that they are looking for regardless of the ordering in which the results are presented.  If this is the case, could we drop these more complex elements for the benefit of making the code more readable and maintainable for the future?  Is there a reason for this that I am trivialising or missing?

File: sso/tests/test_user_api.py [hunk 9]
```diff
@@ -373,3 +374,524 @@ def test_honours_user_permissioms(self, api_client):
...
+        user1 = UserFactory(
+            email='first1.last1@example.com',
+            first_name='First1',
+            last_name='Last1'
+        )
+        user1.groups.add(GroupFactory())
+        oauth_app.users.add(user1)
+
+        api_client.credentials(HTTP_AUTHORIZATION='Bearer ' + token)
+        response = api_client.get(self.GET_USER_SEARCH_URL)
```
Comment (currycoder) [inline sso/tests/test_user_api.py]: Is getting all results with no search query expected/desirable?

File: sso/tests/test_user_api.py [hunk 10]
```diff
@@ -373,3 +374,524 @@ def test_honours_user_permissioms(self, api_client):
...
+        assert len(response.json()) == 2
+
+    @pytest.mark.django_db
+    def test_all_users_scenario_1(self, api_client):
+        """
+        searching with an app that has default_access_allowed True
+        so api lists all users, regardless of app access
+        """
+        search_user, def_oauth_app, token = self.setup_seach_user()
+        user1 = UserFactory(
```
Comment (currycoder) [inline sso/tests/test_user_api.py]: It could be good to set these users up using pytest fixtures.  https://docs.pytest.org/en/latest/fixture.html

File: sso/tests/test_user_api.py [hunk 11]
```diff
@@ -373,3 +374,524 @@ def test_honours_user_permissioms(self, api_client):
...
+        oauth_app.users.add(user1)
+
+        api_client.credentials(HTTP_AUTHORIZATION='Bearer ' + token)
+        response = api_client.get(self.GET_USER_SEARCH_URL)
+
+        assert response.status_code == 200
+        assert len(response.json()) == 2
+
+    @pytest.mark.django_db
+    def test_all_users_scenario_1(self, api_client):
```
Comment (currycoder) [inline sso/tests/test_user_api.py]: The general repeated patterns and structures of `scenario_1`/2/3 here make me think that pytest parameters could be a good win for reducing code.  https://docs.pytest.org/en/latest/parametrize.html

This will allow you to call the same test method with multiple sets of parameters. e.g.

```
@pytest.mark.parametrize(
    'default_access_allowed,expected_results',
    (
         (True, 3),
         (False, 1),
    )
    def test_all_users(self, api_client, default_access_allowed, expected_results):
        ...
        assert len(response.json()) == expected_results
```

File: sso/user/views.py [hunk 12]
```diff
@@ -63,3 +77,77 @@ def retrieve(self, request):
...
+
+    class Meta:
+        model = get_user_model()
+        fields = {
+            'first_name': ('exact', 'icontains'),
+            'last_name': ('exact', 'icontains'),
+        }
+
+
+class UserListViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
```
Comment (currycoder) [inline sso/user/views.py]: I'm a little worried about the lack of queryset limits or pagination here.  Say that we have 20K users and we search for the character `e` - I believe the result set could be pretty massive. Unless I'm missing where pagination is applied?

File: sso/user/views.py (hunk 7 above)
Comment (lgarvey) [reply sso/user/views.py]: We have SamlApplication and OAuthApplication so there's a need to differentiate them.

File: sso/tests/test_user_api.py [hunk 13]
```diff
@@ -894,4 +920,160 @@ def test_all_users_scenario_5(self, api_client):
...
+
+    @pytest.mark.django_db
+    def test_autocomplete_filter_with_fullname_lowercase(self, api_client):
+        search_user, def_oauth_app, token = self.setup_seach_user()
+        user1 = UserFactory(
+            email='first1.last1@example.com',
+            first_name='First1',
+            last_name='Last1'
+        )
+        user1.groups.add(GroupFactory())
```
Comment (lgarvey) [inline sso/tests/test_user_api.py]: Is the user group relevant to the test?

File: sso/tests/test_user_api.py [hunk 14]
```diff
@@ -894,4 +920,160 @@ def test_all_users_scenario_5(self, api_client):
...
+        def_oauth_app.users.add(user1)
+
+        # add a different user to different app
+        oauth_app = ApplicationFactory()
+        user2 = UserFactory(
+            email='first2.last2@example.com',
+            first_name='First2',
+            last_name='Last2'
+        )
+        user2.groups.add(GroupFactory())
```
Comment (lgarvey) [inline sso/tests/test_user_api.py]: Same - wonder if this is important

File: sso/tests/test_user_api.py [hunk 15]
```diff
@@ -894,4 +920,160 @@ def test_all_users_scenario_5(self, api_client):
...
+
+        assert response.status_code == 200
+        assert response.data["count"] == 1
+
+        response = api_client.get(self.GET_USER_SEARCH_URL + '?autocomplete=first2 last2')
+
+        assert response.status_code == 200
+        assert response.data["count"] == 1
+
+    @pytest.mark.django_db
```
Comment (lgarvey) [inline sso/tests/test_user_api.py]: You don't need @pytest.mark.django_db as it's set at the top of the file

File: sso/tests/test_user_api.py [hunk 16]
```diff
@@ -894,4 +920,160 @@ def test_all_users_scenario_5(self, api_client):
...
+        assert response.data["count"] == 3
+
+        response = api_client.get(self.GET_USER_SEARCH_URL + '?autocomplete=First2 Last2')
+
+        assert response.status_code == 200
+        assert response.data["count"] == 1
+
+    @pytest.mark.django_db
+    def test_autocomplete_filter_with_fullname_reverse(self, api_client):
+        search_user, def_oauth_app, token = self.setup_seach_user()
```
Comment (lgarvey) [inline sso/tests/test_user_api.py]: minor - is that a type `seach`?

File: sso/user/views.py (hunk 7 above)
Comment (sekharpanja) [reply sso/user/views.py]: Will see what can be done without overdoing it.

File: sso/tests/test_user_api.py (hunk 10 above)
Comment (sekharpanja) [reply sso/tests/test_user_api.py]: Yep, agreed. Will update the tests.

File: sso/user/views.py (hunk 12 above)
Comment (sekharpanja) [reply sso/user/views.py]: I've added global pagination setting and limit them 100 results atm.

File: sso/tests/test_user_api.py (hunk 13 above)
Comment (sekharpanja) [reply sso/tests/test_user_api.py]: Perhaps not in every test. But it is worth having one test to make sure, adding user to a group will not have adverse affects.

File: sso/tests/test_user_api.py (hunk 15 above)
Comment (sekharpanja) [reply sso/tests/test_user_api.py]: I've missed that bit, will correct them.

File: sso/tests/test_user_api.py (hunk 9 above)
Comment (sekharpanja) [reply sso/tests/test_user_api.py]: Not necessary as per current requirement. But this there might be other apps that might need all users down the line. Results are limited by pagination so that we don't get all users back in one go.
